packages/block-editor/src/store/selectors.js

```diff
@@ -1844,28 +1845,159 @@ export const __experimentalGetAllowedPatterns = createSelector(
+	]
+);
+
+/**
+ * Determines the items that appear in the available pattern transforms list.
+ * There is special handling in two cases:
+ * 1. For some blocks (`blocksToSkip`) when multiple blocks are selected,
+ * don't show any transforms, as it doesn't make sense to try to be too smart.
+ * 2. There are some blocks (`nestedSingleBlocksToHandle`) that makes sense to
+ * replace everything when they are the only block selected.
+ *
+ * For the rest blocks we return a first set of possible eligible block patterns,
+ * by checking the `scope` Patterns API. We still have to recurse through block
+ * pattern's blocks and try to find matches from the selected blocks. Now this
+ * happens in the consumer to avoid heavy operations in the selector.
+ *
+ * @param {Object}  state Editor state.
+ * @param {Object[]} blocks The selected blocks.
+ * @param {?string} rootClientId Optional root client ID of block list.
+ *
+ * @return {WPEditorTransformItem[]} Items that are eligible for a pattern transformation.
+ */
+// TODO tests
+export const __experimentalGetPatternTransformItems = createSelector(
+	( state, blocks, rootClientId = null ) => {
+		if ( ! blocks ) return EMPTY_ARRAY;
+		// There are some blocks like `Template Part` that makes sense
+		// to replace everything when they are the only block selected.
+		const isSingleBlockSelected = blocks.length === 1;
+		const nestedSingleBlocksToHandle = [ 'core/template-part' ];
+		if (
+			isSingleBlockSelected &&
+			nestedSingleBlocksToHandle.includes( blocks[ 0 ].name )
+		) {
+			const [ selectedBlock ] = blocks;
+			let { name: blockName } = selectedBlock;
+			// Try to find an active block variation for `Template Part`,
+			// to show patterns with `scope` including the block variation
+			// name (ex. `core/template-part/{variation name}). If no active
+			// block variation is found, we search for patterns for the base
+			// block.
+			const variations = getBlockVariations( blockName );
+			if ( variations?.length ) {
+				const match = variations.find( ( variation ) =>
+					variation.isActive?.(
+						selectedBlock.attributes,
+						variation.attributes
+					)
+				);
+				if ( match ) {
+					blockName = `${ blockName }/${ match.name }`;
+				}
+			}
+			return __experimentalGetScopedBlockPatterns(
+				state,
+				'transform',
+				blockName
+			);
+		}
+
+		// Create a Set of the selected block names that is used in patterns filtering.
+		const selectedBlockNames = Array.from(
+			blocks.reduce( ( accumulator, block ) => {
+				accumulator.add( block.name );
+				return accumulator;
+			}, new Set() )
+		);
+
+		// If a selected block is nested (with InnerBlocks like Group or Columns)
+		// return an EMPTY_ARRAY, as it doesn't make sense to try to be too smart.
+		// In such blocks a user could have anything inside as content, so don't show
+		// any transforms. In these blocks it only makes sense to show `block` scoped
+		// patterns during insertion, where no content exists yet.
```

Comment:
The difference between such blocks is that some of them are mean more as `grouping` blocks. Examples are `Group`, `Columns`, `Template Part` and:
>// In such blocks a user could have anything inside as content, so don't show
		// any transforms. In these blocks it only makes sense to show `block` scoped
		// patterns during insertion, where no content exists yet.

This could work differently in simpler(more semantic) blocks with InnerBlocks like `Buttons`, `Quote`, `Cover`. While the user could still have anything inside it's still more targeted to specific semantics and the implementation at this point kept all the selected block's InnerBlocks.

This is a great point that I wasn't sure and wanted feedback and maybe it should be handled not by the framework but instead be explicitly declared in the block somewhere(?)